[Post] Android Navigation Component - Expectations, Conclusions & Tips - Part 1 (#116)
xmartlabs/blog: _posts/2020-05-12-Android-Nav-Component.md

---
layout: post
title: "Android Navigation Component - Expectations, Conclusions & Tips - Part 1"
date: 2020-05-12 10:00:00
categories: android, architecture components, jetpack, navigation component
author_id: mirland
show: true
featured_image: /images/android_navigation_blog_part_one/navigation_banner.jpg
crosspost_to_medium: false
---
<!--- STOPSHIP: Change date --->

This year the [Google I/O](https://events.google.com/io/) conference was canceled, so I think it's a good time to talk about one of [Jetpack's](https://developer.android.com/jetpack) biggest Architecture component introduced last year, the [Android Navigation Component](https://developer.android.com/guide/navigation).

The aim of this series of posts is to talk about two important items that will help you decide on whether to use this library or not:
1. The expectations and conclusions of using it for more than 9 months.
1. Some tips and helper classes that are very helpful if you do want to start using it (next post). 

## What were we expecting from a new navigation library?
That was the first question that came to my mind when we first thought about using it.

> There are a bunch of navigation libraries in existence for Android, and at that moment we already had a solid navigation architecture based on the Router pattern, so, what were we expecting of it?

Reviewer: ```suggestion
There are a bunch of navigation libraries in existence for Android, and at that moment we already had a solid navigation architecture based on the Router pattern, so, what were we expecting from it?
```

- The ability to define **nested graphs**, which are graphs that will be included by other graphs.
So that you can define a specific app flow part in a nested graph to work with small diagrams and be able to handle complex app flow.

> The login is a common example for both items: suppose that the login epic is a complex flow that contains a login screen, a register screen, and a terms and conditions screen.
If you add all of these screens to your main graph, it'll make it grow too fast -risking its future maintainability- so I recommend including a login subgraph instead.
Using this approach I found some advantages: the main graph is easier to understand and you can also re-utilize a given flow in multiple places in the app.
We know that re-utilization is an excellent pattern: if you have to change something, you will do it just once.

One disadvantage we found is related to how Android Studio stores the graph nodes position.
These are saved in a huge and complex `.idea/navEditor.xml` file.
If you want to save the positions of all graphs nodes -in order for the graph to be consistent among your team, for instance- you need to track this file and include it in your repo.

> Sometimes this file will be randomically corrupted and you'll lose the positions of all nodes.

Reviewer: ```suggestion
Sometimes this file will be randomly corrupted and you'll lose the positions of all nodes.
```

If that is your case, you can check out the last valid file and it'll be fixed.
Moreover this file suffers from an awful lot of git conflicts, so it's a bit tedious to maintain it.

As I commented before, you can [share data between the nodes (screens) using arguments](https://developer.android.com/guide/navigation/navigation-pass-data).
The navigation component includes a Gradle plugin that has a pretty self-explanatory name: [Safe Args](https://developer.android.com/guide/navigation/navigation-pass-data#Safe-args).
That's really good because you can share arguments in a safe way and you don't have to care about the ugly key-value bundle used up until now.

Although I mentioned that sharing data in that way is good, we found that sharing data among nested graphs is not safe yet, since we have to declare these arguments and if we forget the app crashes in runtime.
This is something likely to improve in future version.

### Debug and test your app

> The navigation component is also a very useful asset when it comes to **debugging** and testing work in progress (WIP) features easily in your app.

Reviewer: ```suggestion
The navigation component is also a very useful asset when it comes to **debugging** and testing work-in-progress (WIP) features easily in your app.
```

Because it's being used like an adjective.

It let us start the app on a particular screen and this is useful during development because we don't even need to navigate manually.
In order to launch a particular screen we change the `startDestination` property and set up its default arguments in the navigation graph if needed.
This way, when you run the app for the next time, this internal screen will be opened right away, awesome right?
On the other hand, debugging the graph status is a bit hard, not being able to know what's the current graph path.
To solve this issue we created a cool tool to log the current path, which I'll introduce in the next post.

The library includes a [**testing** module](https://developer.android.com/guide/navigation/navigation-testing), which lends a helping hand when testing your app's navigation logic.
Although I didn't get to use it a lot, I found it useful and easy to use whenever I had to.

> ### Jetpack & Android Architecture Componentsponent

Reviewer: ```suggestion
### Jetpack & Android Architecture Components
```

It seems like a bug from Grammarly.

---
layout: post
title: "Android Navigation Component - Expectations, Conclusions & Tips - Part 1"
date: 2020-05-12 10:00:00
categories: android, architecture components, jetpack, navigation component
author_id: mirland
show: true
featured_image: /images/tflite_coreml/featured.png
---
<!--- STOPSHIP: Change date and featured_image --->

This year the [Google I/O](https://events.google.com/io/) conference was canceled, so I think it's a good time to talk a about one of [Jetpack's](https://developer.android.com/jetpack)  biggest Architecture component introduced last year, the [Android Navigation Component](https://developer.android.com/guide/navigation).

The aim of this series of posts is to talk about two important items that will help you decide on whether to use this library or not:
1. The expectations and conclusions of using it for more than 9 months.

> 1. Some tips and helper classes that are very helpful if you do want to start using it (next post). 

Reviewer 2: ```suggestion
2. Some tips and helper classes that are very helpful if you do want to start using it (next post). 
```

Author: It's better if we keep `1.`. It's displayed [`2`](https://blog-staging.xmartlabs.com/2020/05/12/Android-Nav-Component/) but if you add a new item it's not needed to increase it

- A library which could provide us with a better understanding of the app's main flows and features. 
- Share data between views easily.
- A good integration between all of Jetpack's Components.

## What did we find?

The **integration** is very easy, you can follow [Android's documentation](https://developer.android.com/guide/navigation/navigation-getting-started) and check the [sample projects](https://github.com/android/architecture-components-samples) upon doing it.

After the integration, you'll have to define the app's main **navigation graph**, one of the most important components of this library.

<!--- STOPSHIP: Add image --->
[IMAGE](TODO)

The graph is stored in a simple XML file, where the nodes are the app's screens (fragments, dialogs, etc) and the actions are the edges that allow for navigating between nodes.
Each node is able to declare arguments, which ends up being data shared within the navigation flow.

I found the nav graph extremely useful, with two key advantages:
- You can see the whole app flow represented as a navigation graph.
It's useful not only for you but especially for new colleagues joining your project. 

> - The ability to define nested graphs, so that you can have big epics represented in nested graphs.

Reviewer 2: i didn't get this i think that the benefits of nested graphs are that you can include a flow into another graf without increasing it size significantly i think that maybe puting something more in that line would be more useful like: 

```
The ability to define nested graphs, so that you can include an epic into a graph without increasing its size or nodes significantly.
```

> - The ability to define nested graphs, so that you can have big epics represented in nested graphs.

Author: ```suggestion
- The ability to define nested graphs, so that you can include an epic into a graph without increasing its size or nodes significantly.
```

The login is a common example: suppose that the login epic contains a login screen, a register screen and a terms and conditions screen.
If you add all of these screens to your main graph, it'll make it grow too fast -risking its future maintainability- so I recommend including a login subgraph instead.
Using this approach I found some advantages: the main graph is easier to understand and you can also re-utilize a given flow in multiple places in the app.
We know that reutilization is an excellent pattern: if you have to change something, you will do it just once.

One disadvantage we found is related to how Android Studio stores the graph nodes position.
These are saved in a huge and complex `.idea/navEditor.xml` file.
If you want to save the positions of all graphs nodes -in order for the graph to be consistent among your team, for instance- you need to track this file and include it in your repo.

> Sometimes this file will be corrupted and you'll lose the positions of all nodes.

Author: Not sure, AS issues. I think it's when multiple users are working in new graphs, but I'm not sure

Moreover this file suffers of an awful lot of git conflicts, so it's a bit tedious to maintain it.

As I commented before, you can [share data between the nodes (screens) using arguments](https://developer.android.com/guide/navigation/navigation-pass-data).
The navigation component includes a Gradle plugin which has a pretty self-explanatory name: [Safe Args](https://developer.android.com/guide/navigation/navigation-pass-data#Safe-args).
That's really good because you can share arguments in a safe way and you don't have to care about the ugly key-value bundle used up until now.

Although I mentioned that sharing data in that way is good, we found that sharing data between different nested graphs is neither good nor safe.

> If you want to do that, you'll have to declare the arguments manually in the graph's XML file, and if you ever happen to forget it, the app will end up crashing.

Author: The safe args is a plugin should do that, but right now it's not working fine in this case

### Debug and test your app

The navigation component is also a very useful asset when it comes to **debugging** and testing work in progress (WIP) features easily in your app.
We found an interesting approach that proved to work very well for us: 
Suppose you are working on a feature in some internal screen.
When you run the application you have to navigate through the app to open the screen and test your progress in that feature.
We realized that we were wasting a lot of time navigating inside the app to test how our new internal screen looks.

> If you use this library, you can change the `startDestination` property to be your WIP screen.

Author: What I do here is define it as a default parameters. (I don't share entities, I use to shere type, ids and that stuff, so I can do it easily). Maybe I can comment it

This way when you run the app next time, this internal screen is opened right away, awesome right?
On the other hand, debugging the graph status is a bit hard, not being able to know what's the current graph path.
To solve this issue we created a cool tool to log the current path, which I'll introduce in the next post.

The library includes a [**testing** module](https://developer.android.com/guide/navigation/navigation-testing), which lends a helping hand when testing your app's navigation logic.
Although I didn't get to use it a lot, I found it useful and easy to use whenever I had to.

### Jetpack & Android Arch Component

I will now talk about one of the most important points of this library.

> The integration between this and the other [Android Architecture Components](https://developer.android.com/guide/navigation/navigation-deep-link).

Author: haha np

As you may know, some years ago Google released the Android Arch Components, a collection of libraries that help you design robust, testable, and maintainable apps.

> In this post I will not delve into them, but there are some of these components that are very important in all new Android Applications, such as the [ViewModel](https://developer.android.com/topic/libraries/architecture/viewmodel).

Reviewer 2: ```suggestion
In this post I will not dive into them, but there are some of these components that are very important in all new Android Applications, such as the [ViewModel](https://developer.android.com/topic/libraries/architecture/viewmodel).
```
seems that dive is more common that delve

In my opinion, if you combine all the Android Arch Components they always work like a charm, and this one is not the exception.

All graphs have an associated [lifecycle scope](https://developer.android.com/topic/libraries/architecture/lifecycle), so that you can create ViewModels associated to the graph's scope.

> That means that you can share data through the screens using a ViewModel that's in turn associated to a graph.

Reviewer 2: ```suggestion
That means that you can share data through the screens using a ViewModel that's associated to a graph.
```
